portal/index.php

```diff
@@ -27,15 +27,52 @@
 
 //includes
 require_once('../interface/globals.php');
+require_once(dirname(__FILE__) . "/lib/appsql.class.php");
+$logit = new ApplicationTable();
 
+use OpenEMR\Common\Crypto\CryptoGen;
 use OpenEMR\Core\Header;
 
 //exit if portal is turned off
 if (!(isset($GLOBALS['portal_onsite_two_enable'])) || !($GLOBALS['portal_onsite_two_enable'])) {
     echo xlt('Patient Portal is turned off');
     exit;
 }
-
+$auth['portal_pwd'] = '';
+if (isset($_GET['woops'])) {
+    unset($_GET['woops']);
+    unset($_SESSION['password_update']);
+}
+if (isset($_GET['forward'])) {
+    $auth = false;
+    if (strlen($_GET['forward']) >= 64) {
+        $crypto = new CryptoGen();
+        $one_time = $crypto->decryptStandard($_GET['forward'], '', 'database');
+        $auth = sqlQueryNoLog("Select * From patient_access_onsite Where portal_login_username Like BINARY ?", array(($one_time . '%')));
```

Comment:
for me:
This storing of authentication tokens in the username field is giving me palpitations :)
I just want us to avoid another hall of shame in the future:
https://community.open-emr.org/t/openemr-security/10597
(I am picturing myself trying to explain to journalist why we did this :) )
The code here seems solid like you have it (unless somebody decides to make a long username for themselves with a hex representation of a date at the end :) ), but what happens when we die off (or even just after a couple weeks when we forget about this and move on to something else like FIHR) and somebody does a harmless improvement that unwittingly creates a potentially nasty vulnerability. I think the only benefit this is providing is the token automatically goes away after it's used. I am happy to add this issue to my list (I think it will just take changes in a couple lines of code).

Reply:
I could argue this all the way to the Supreme Court but, your point about future modifications is completely valid. 
If I hadn't promise this to several people i'd of pulled it and waited for Fhir authentication to be done and used it. Still we got away with existing security too long and something needed to be done as stop gap.
I do promise. This will be the last time I fool with passwords......

I'll change while I have project queued up and fix my logic nightmare. Normally wouldn't do it that way but now you've got so paranoid and seeing boogie men everywhere!

Reply:
Still, i'm glad I did the pin.

Reply:
Agreed on the pin :)